cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -1008,3 +1026,73 @@ func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath stri
 	}
 	return ht.c.Modify(ctx, modification)
 }
+
+// publishNotifications monitors the containers notification queue.
+// Currently only OOM events are published.
+//
+// This MUST be called via a goroutine to wait on a background thread.
+// Returned error is used for scoping, and not meant to be consumed
+func (ht *hcsTask) publishNotifications(ctx context.Context) {
+	ctx, span := trace.StartSpan(ctx, "hcsTask::publishNotifications")
+	defer span.End()
+	var err error
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.TaskID, ht.id))
+	entity := log.G(ctx).WithField(logfields.TaskID, ht.id)
+
+	var nq *queue.MessageQueue
+	nq, err = ht.c.Notifications()
+	if err != nil {
+		// includes `notifications.ErrNotificationsNotSupported`, where container
+		// does not publish any notifications
+		entity.WithError(err).Warning("unable to retrieve container notifications channel")
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			var v interface{}
+			v, err = nq.ReadOrWait()
+			if err != nil {
+				if err == queue.ErrQueueClosed {
+					// set err to nil for oc.SetSpanStatus
+					err = nil
+				} else {
+					// this could be a transient fault, but to prevent infinite
+					// errors and polluting the log stream, break here as well
+					entity.WithError(err).Warning("could not read from notification queue")
+				}
+
+				return
+			}
+
+			ntf, ok := (v).(notifications.Message)
+			if !ok {
+				err = fmt.Errorf("expected type of 'notifications.Notification{}', received %T", v)
```

Comment:
*notifications.Message

Reply:
Also, do we want to return here?

Reply:
I didnt think a bad message should bring down the queue, but I can add it in if you think its a good idea

Reply:
Oh I see, we could just `continue` then. It's just weird to me that if this cast is not okay, we'd log twice: once here and once below in the default switch case.

Reply:
done